tracker/internal/graphstore/v1beta/server.go

```diff
@@ -0,0 +1,263 @@
+		Valid: true,
+	}
+
+	for _, edge := range request.GetEdges() {
+		item := ConvertEdge(edge, EncodingProtocolBuffers)
+		item.DateDeleted = dateDeleted
+		items = append(items, item)
+	}
+
+	for _, node := range request.GetNodes() {
+		item := ConvertNode(node, EncodingProtocolBuffers)
+		item.DateDeleted = dateDeleted
+		items = append(items, item)
+	}
+
+	if err := s.Driver.Delete(ctx, items); err != nil {
+		return nil, err
+	}
+
+	return &graphstore.DeleteResponse{}, nil
+}
+
+func paginate(lastToken string, pageSize int) (int, string) {
+	offset := 0
+
+	if lastToken != "" {
+		data, err := base32.StdEncoding.DecodeString(lastToken)
+		if err != nil {
+			return offset, ""
+		}
+		offset = int(binary.BigEndian.Uint64(data))
+	}
+
+	nextOffset := offset + pageSize
+
+	nextToken := make([]byte, 8)
+	binary.BigEndian.PutUint64(nextToken, uint64(nextOffset))
+
+	return offset, base32.StdEncoding.EncodeToString(nextToken)
+}
+
+// List ...
+func (s *GraphStoreServer) List(ctx context.Context, request *graphstore.ListRequest) (*graphstore.ListResponse, error) {
+	limit := int(request.GetPageSize())
+
+	offset, nextPageToken := paginate(request.GetPageToken(), limit)
+
+	results, hasNext, err := s.Driver.List(ctx, request.GetKind(), offset, limit)
+	if err != nil {
+		return nil, err
+	}
+
+	if !hasNext {
+		nextPageToken = ""
+	}
+
+	nodes := make([]*graphstore.Node, 0)
+	edges := make([]*graphstore.Edge, 0)
+
+	for _, item := range results {
+		if item == nil {
+			break
+		}
+
+		node, edge, err := ConvertGraphData(item)
+		if err != nil {
+			logrus.Errorf("encountered issue decoding row [%s, %s, %s] in database: %v",
+				string(item.K1), string(item.K2), string(item.K3), err)
+			continue
+		}
+
+		if node != nil {
+			nodes = append(nodes, node)
+		} else {
+			edges = append(edges, edge)
+		}
+	}
+
+	return &graphstore.ListResponse{
+		NextPageToken: nextPageToken,
+		Nodes:         nodes,
+		Edges:         edges,
+	}, nil
+}
+
+func processNeighbors(neighborData []*GraphData, keyFn func(edge *graphstore.Edge) string) []*graphstore.Neighbor {
+	results := make([]*graphstore.Neighbor, 0)
+	index := make(map[string]int)
+
+	for _, neighbor := range neighborData {
+		node, edge, err := ConvertGraphData(neighbor)
+		if err != nil {
+			logrus.Errorf("encountered issue decoding row [%s, %s, %s] in database: %v",
+				string(neighbor.K1), string(neighbor.K2), string(neighbor.K3), err)
+			continue
+		}
+
+		if node != nil {
+			key := string(node.GetKey())
+			if _, ok := index[key]; !ok {
+				pos := len(results)
+				results = append(results, &graphstore.Neighbor{
+					Node:  node,
+					Edges: make([]*graphstore.Edge, 0),
+				})
+				index[key] = pos
+			}
+		} else {
+			pos := index[keyFn(edge)]
+			results[pos].Edges = append(results[pos].Edges, edge)
+		}
+	}
+
+	return results
+}
+
+// Neighbors ...
+func (s *GraphStoreServer) Neighbors(ctx context.Context, request *graphstore.NeighborsRequest) (*graphstore.NeighborsResponse, error) {
```

Comment:
Is this the equivalent of `FindUpstream` / `FindDownstream` in alpha's implementation?

Also, how does it work if I need only the "dependent" nodes or the "dependency" nodes?

Reply:
Upstream/Downstream were hangovers from the previous implementation. Most people familiar working with graphs expect a Neighbors call. This call can be done 3 different ways: 

1. by passing in a `from` node, you will get all upstream (neighbors from node, implies egressing edges)
1. by passing in a `to` node, you will get all downstream (neighbors pointing to node, implies ingressing edges)
1. by passing in a `node` you will get _ALL_ neighbors

Most graph interfaces don't let you freely invert relationships so they don't need this kind of complexity.